soteria/lib/symex/tree_block.ml

```diff
@@ -0,0 +1,674 @@
+      | Owned left, Owned right ->
+          let v, children = MemVal.merge ~left ~right in
+          (Owned v, children)
+
+    let split ~at node =
+      match node with
+      | Owned v ->
+          let lift = map_tree (fun v -> Owned v) in
+          let+ left, right = MemVal.split ~at v in
+          (lift left, lift right)
+      | NotOwned Totally ->
+          return (Leaf (NotOwned Totally), Leaf (NotOwned Totally))
+      | NotOwned Partially -> failwith "Should never split an intermediate node"
+  end
+
+  module Tree = struct
+    type t = (MemVal.t, sint) tree
+
+    let rec pp ft { node; range; children } =
+      let open Fmt in
+      pf ft "@[<v 2>{%a %a%a}@]" Node.pp node Range.pp range
+        (option ~none:nop (fun fmt (l, r) ->
+             pf fmt " -->@,@[%a@,%a@]" pp l pp r))
+        children
+
+    let make = make_tree
+    let is_empty t = Node.is_empty t.node
+    let not_owned range = make ~node:(NotOwned Totally) ~range ?children:None ()
+
+    let rec map_leaves t f =
+      match t.children with
+      | None -> f t
+      | Some (l, r) ->
+          let** l = map_leaves l f in
+          let++ r = map_leaves r f in
+          { t with children = Some (l, r) }
+
+    let rec iter_leaves_rev t f =
+      match t.children with
+      | None -> f t
+      | Some (l, r) ->
+          iter_leaves_rev r f;
+          iter_leaves_rev l f
+
+    let iter_leaves_offsets t =
+      Iter.map
+        (fun leaf ->
+          let offset, _ = leaf.range in
+          let offset = offset -@ fst t.range in
+          (leaf.node, offset))
+        (iter_leaves_rev t)
+
+    let of_children_s ~left ~right =
+      let range = (fst left.range, snd right.range) in
+      let node, children = Node.merge ~left:left.node ~right:right.node in
+      let children =
+        match children with
+        | `Complete -> None
+        | `KeepChildren -> Some (left, right)
+      in
+      return { range; children; node }
+
+    let of_children _ ~left ~right = of_children_s ~left ~right
+
+    let with_children t ~left ~right =
+      return { t with children = Some (left, right) }
+
+    let rec offset ~by tree =
+      let low, high = tree.range in
+      let range = (low +@ by, high +@ by) in
+      let children =
+        Option.map (fun (l, r) -> (offset ~by l, offset ~by r)) tree.children
+      in
+      make ~node:tree.node ~range ?children ()
+
+    let rec tree_of_rec_node range = function
+      | Leaf node -> return @@ make ~node ~range ()
+      | Node (left, at, right) ->
+          let left_span, right_span = Range.split_at range (fst range +@ at) in
+          let* left = tree_of_rec_node left_span left in
+          let* right = tree_of_rec_node right_span right in
+          of_children_s ~left ~right
+
+    let rec split ~range t : (Node.t * t * t) Symex.t =
+      (* this function splits a tree and returns the node in the given range *)
+      (* We're assuming that range is inside old_span *)
```

Comment:
```suggestion
    (** [split ~range t] isolates [range] from [t]. Precondition: [range] is a
        strict subrange of [t.range] (neither empty nor equal to [t.range]).
        Returns [(node, left, right)] where:
        - [node] is the node covering exactly [range]
        - [left] and [right] can be safely set as [t]'s children, and [range]
          lies within either [left] or [right].

        If [range] touches the left or right edge of [t], a single split
        suffices. Otherwise, we first split at [fst range] to peel off the left
        part, then carve [range] out of the resulting right part. This makes the
        procedure right-biased (it prefers introducing structure on the right)
    *)
    let rec split ~range t : (Node.t * t * t) Symex.t =
```